tests/python_functional/src/syslog_ng_config/statements/destinations/destination_driver.py

```diff
@@ -20,15 +20,40 @@
 # COPYING for details.
 #
 #############################################################################
+from src.syslog_ng_config.statements.config_statement import ConfigStatement
+from src.syslog_ng_config.statements.destinations.destination_reader import DestinationReader
 
 
-class DestinationDriver(object):
+class DestinationDriver(ConfigStatement):
     group_type = "destination"
 
-    def __init__(self, positional_parameters=None, options=None):
+    def __init__(self, positional_parameters=None, options=None, driver_io_cls=None, line_parser_cls=None):
         if positional_parameters is None:
             positional_parameters = []
         self.positional_parameters = positional_parameters
         if options is None:
             options = {}
         self.options = options
+
+        self.driver_io_cls = driver_io_cls
+        self.line_parser_cls = line_parser_cls
+        self.destination_reader = None
+        self.init_destination_reader()
+        super(DestinationDriver, self).__init__()
+
+    def init_destination_reader(self):
+        if self.driver_io_cls and self.line_parser_cls and self.positional_parameters:
+            self.destination_reader = DestinationReader(self.driver_io_cls, self.line_parser_cls)
```

Comment:
We discussed IRL that it would worth separating the IO related logic (DriverIO) and the Config logic (Driver).

Now a driver-io is embedded into the driver. This causes some problem: their lifecycle / management differs. For example, configuration is applied during start and reload. However an IO, like http server may or may not need to be started right away. Or in a FileSource scenario, we might not want to create the file right away.

These two things might worth store and manage separately.

For example (more like a pseudocode):
```
http = HttpDestination(port=5555)

generator = generator_source(num=1)
http = HttpDestination(port=5555)
syslog_ng.config_add(generator, http)

simulator = HttpSimulator()
simutor.init(parameters)
simulator.start()

syslog_ng.start()
time.sleep(1)
syslog_ng.stop()

simulator.stop()
```

Reply:
Yes, I have already started this implementation and it seems it will work in our cases.
But this concept will broke our existing user API. Before going ahead I would like to discuss it IRL. So that's why I put back WIP flag